### PR TITLE
expression: rewrite builtin string function: LPAD, RPAD

### PR DESCRIPTION
**Pull request by zz-jason**
to https://github.com/pingcap/tidb/issues/4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1345,56 +1347,198 @@ func trimRight(str, remstr string) string {
...
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if types.IsBinaryStr(args[0].GetType()) || types.IsBinaryStr(args[2].GetType()) {
+		types.SetBinChsClnFlag(bf.tp)
+		sig := &builtinLpadBinarySig{baseStringBuiltinFunc{bf}}
+		return sig.setSelf(sig), nil
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: should we check `mysql.HasBinaryFlag(args[i].GetType().Flag)` here,
and set bf.tp.Flag |= mysql.BinaryFlag if it's true

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1345,56 +1347,198 @@ func trimRight(str, remstr string) string {
...
+	}
+	sig := &builtinLpadSig{baseStringBuiltinFunc{bf}}
+	return sig.setSelf(sig), nil
+}
+
+type builtinLpadBinarySig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString evals LPAD(str,len,padstr)).
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: unnecessary `)`

File: expression/builtin_string.go [hunk 3]
```diff
@@ -1345,56 +1347,198 @@ func trimRight(str, remstr string) string {
...
+		str = strings.Repeat(padStr, repeatCount)[:tailLen] + str
+	}
+	return str[:targetLength], false, nil
+}
+
+type builtinLpadSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString evals LPAD(str,len,padstr)).
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 4]
```diff
@@ -1345,56 +1347,198 @@ func trimRight(str, remstr string) string {
...
-	sig := &builtinRpadSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if types.IsBinaryStr(args[0].GetType()) || types.IsBinaryStr(args[2].GetType()) {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/integration_test.go [hunk 5]
```diff
@@ -86,6 +86,30 @@ func (s *testIntegrationSuite) TestFuncREPEAT(c *C) {
 	r.Check(testkit.Rows("<nil> <nil> <nil> <nil> <nil> <nil>"))
 }
 
+func (s *testIntegrationSuite) TestFuncPad(c *C) {
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: TestFuncLpadAndRpad be more explicit?

File: expression/typeinferer_test.go [hunk 6]
```diff
@@ -255,7 +255,11 @@ func (ts *testTypeInferrerSuite) TestInferType(c *C) {
 		{"unhex(12)", mysql.TypeVarString, charset.CharsetUTF8, 0},
 		{"DATE_FORMAT('2009-10-04 22:23:00', '%W %M %Y')", mysql.TypeVarString, charset.CharsetUTF8, 0},
 		{"rpad('TiDB', 12, 'go')", mysql.TypeVarString, charset.CharsetUTF8, 0},
+		{"rpad(c_binary, 12, 'go')", mysql.TypeVarString, charset.CharsetBin, 0},
```
Comment (XuHuaiyu) [inline expression/typeinferer_test.go]: we add test in plan/typeinfer_test.go
not expression/typeinferer_test.go

**Comment by XuHuaiyu** [comment]
please address the comments

**Comment by zz-jason** [comment]
@XuHuaiyu adressed, PTAL again.

**Comment by zz-jason** [comment]
@breeswish @jackysp @winkyao PTAL

**Comment by zz-jason** [comment]
@hanfei1991 PTAL

File: expression/builtin_string.go [hunk 7]
```diff
@@ -1470,56 +1473,225 @@ func trimRight(str, remstr string) string {
 	}
 }
 
-type rpadFunctionClass struct {
+func getFlen4LpadAndRpad(sc *variable.StatementContext, arg Expression) (int, error) {
+	if constant, ok := arg.(*Constant); ok {
+		length, isNull, err := constant.EvalInt(nil, sc)
+		if isNull || err != nil || length > mysql.MaxBlobWidth {
+			return mysql.MaxBlobWidth, errors.Trace(err)
```
Comment (hanfei1991) [inline expression/builtin_string.go]: if isNull is true, we still set max width ?

Comment (hanfei1991) [inline expression/builtin_string.go]: If err is not nil, we should not raise an error, because this function may be not executed in fact.

Comment (zz-jason) [inline expression/builtin_string.go]: 1. yes, if is null is true, we still set max width, to be compatible with mysql
2. fixed

**Comment by zz-jason** [comment]
updated, @hanfei1991 PTAL

File: expression/builtin_string.go [hunk 8]
```diff
@@ -1470,56 +1473,222 @@ func trimRight(str, remstr string) string {
 	}
 }
 
-type rpadFunctionClass struct {
+func getFlen4LpadAndRpad(sc *variable.StatementContext, arg Expression) int {
+	if constant, ok := arg.(*Constant); ok {
+		length, isNull, err := constant.EvalInt(nil, sc)
+		if isNull || err != nil || length > mysql.MaxBlobWidth {
```
Comment (hanfei1991) [inline expression/builtin_string.go]: Log the error

**Comment by zz-jason** [comment]
updated, @hanfei1991 @lamxTyler PTAL again.

**Pull request merged**